forcing the ConsoleDestination to use the main queue

mishagray opened the pull request:
to avoid print() issues..

About half the time I am seeing ZERO output in my debugger for stuff going to the console.

I think there are issues with using Swift 'print' outside of the mainq.

This change just forces the ConsoleDestionation to use the mainq as it's queue.   
This seems to have gotten rid of my 'where are my console logs' problem.

 Also allows people to define their own custom queues for their own Destinations.

davidrothera commented:
Hmm I haven't seen any issues with console logging at all. Are you running the command in something that is about to kill itself (like a Swift cli tool)?

mishagray commented:
No.. this is happening in my debugger... I get absolutely ZERO output.  Nothing is hitting the console at all.   My best guess is that some other piece of code is ALSO calling print() at the same time from the mainq, while SwiftBeaver is calling it from a background queue, which seems to screw up the console output completely.   This may also be new in XCode 7.3/swift 2.2  (Before I would get garbled print() characters from different threads).

If you have all your output via SwifyBeaver you won't see this issue.  But I kind of assume I won't be the last person to have conflicts with print(), especially since it is being called from a background thread.

There seem to be issues with print() not being thread-safe: https://forums.developer.apple.com/thread/21949
http://stackoverflow.com/questions/28753884/didreceiveremotenotificationfetchcompletionhandler-but-the-completion-handler/28754349#28754349

The OTHER answer, seems to be just to replace print() with NSLog(), which IS thread-safe.

mishagray commented:
Theory:  Swift 2.2 standard libs are now partially 'linux' compatible, so maybe they changed something about thread-safety when they added support for Linux to swift 2.2.   I have no evidence for this.. but it would make sense why it is happening now.

davidrothera commented:
Hmm that sucks, yeah come to think of it once my controllers have loaded all logging is from SB so probably not running into this.

skreutzberger commented:
Thanks for reporting that print may not be thread-safe. It would be great if we could find a simpler solution which fixes potential issues.

Did you try the recently introduced asynchronously flag set to false? 

It is used in the code [here](https://github.com/SwiftyBeaver/SwiftyBeaver/blob/master/sources/SwiftyBeaver.swift#L130). 

Also what about NSLog? I am using it for internally testing SwiftyBeaver because I also do not trust print(). Maybe we can fully migrate away from print() to NSLog?

mishagray commented:
Yeah. ... The more I think about it ... I think NSLog is better than using the main queue.  Cause even using the Main queue doesn't prevent some code using print in some background queue, and still causing issues.  Plus it loads up the Main q with some extra logging duties that maybe we don't want to require. 

I can change it and verify that tomorrow.    

Of course using dispatch_sync vs dispatch_async blocks the calling thread until the block is done running, which can potentially cause other strange deadlocks if you aren't careful.  However GCD is usually smart about dispatch_sync when done from mainQ to mainQ.  But mixing other serial queues can be dangerous.  

But it would make sure that a log message is written before the log operation itself exists.  Which can be useful for debugging sometimes.    Our other option is to add support for "Immediate" mode, which would have the destination use the calling threads execution context.  (I.e. Don't dispatch at all).  Which can also be useful in making sure a log message is written before the log operation is done ... But could mean log messages get written out of order.  And I think would be bad for destinations like files, etc.  But might be perfect for console logging.  

But if print() isn't thread safe ... None of them eliminate the the problem.  NSLog I think is our best bet.

davidrothera commented:
Sounds like a good idea, we just need to ensure that NSLog is a first class citizen in the Open Source StandardLib so that SPM still works fine :fireworks:

mishagray commented:
Looks like it. 

https://github.com/apple/swift-corelibs-foundation/blob/3579f1f306182e4de48a35dfd9067eff22cee27a/Foundation/NSLog.swift

skreutzberger commented:
@mishagray Could you develop it?

mishagray commented:
So I added a bunch of features in my branch... party cause they I think they were mild improvement... but mostly cause I was chasing this problem.

So I made the executionContext configurable (async/sync/custom queues, custom sinks)...
I added support to select print or NSLog or add a custom output...
Added support to print using 'debugDescription' instead of 'description' (which I found only mildly useful, since most of the time people use interpolated strings... and it doesn't work with that).

And I added a small optimization where you only 'build' the log string once, even if you have multiple destinations (and never build it if you have no enabled destinations).


BUT - None of them fixed my issue.   I think I have tracked the real culprit to Xcode 7.3 debugger.  Which seems to have bugs.    Downloading 7.3.1 GM right now to see if it fixes things.

mishagray commented:
I pushed my stuff here... We should probably close this PR.
https://github.com/SwiftyBeaver/SwiftyBeaver/pull/89
